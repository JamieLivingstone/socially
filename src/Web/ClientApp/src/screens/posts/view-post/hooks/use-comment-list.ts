import axios from 'axios';
import { useInfiniteQuery } from 'react-query';

type CommentList = {
  hasNextPage: boolean;
  hasPreviousPage: false;
  items: Array<{
    author: {
      name: string;
      username: string;
    };
    createdAt: string;
    id: number;
    message: string;
  }>;
  pageNumber: number;
  totalCount: number;
  totalPages: number;
};

export function useCommentList(slug: string) {
  const { data, hasNextPage, fetchNextPage, isFetchingNextPage } = useInfiniteQuery(
    ['comments', slug],
    async function fetchCommentListRequest({ pageParam = 1 }) {
      const { data } = await axios.get<CommentList>(
        `/api/v1/posts/${slug}/comments?pageNumber=${pageParam}&pageSize=10`,
      );

      return data;
    },
    {
      getNextPageParam: (lastPage, pages) => {
        return lastPage.hasNextPage ? pages.length + 1 : false;
      },
      suspense: false,
    },
  );

  return {
    pages: data?.pages ?? [],
    hasNextPage: hasNextPage,
    fetchNextPage,
    isFetchingNextPage,
  };
}
