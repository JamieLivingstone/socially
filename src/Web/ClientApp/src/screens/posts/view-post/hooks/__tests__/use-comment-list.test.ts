import { renderHook } from '@testing-library/react-hooks';
import moxios from 'moxios';
import { expect } from 'vitest';

import Providers from '@providers';

import { CommentList, useCommentList } from '../use-comment-list';

describe('useCommentList', () => {
  test('returns paginated comment list', async () => {
    const response: CommentList = {
      hasNextPage: false,
      hasPreviousPage: false,
      items: [
        {
          id: 1,
          message: 'Comment one',
          createdAt: '2022-03-14T08:28:30.959Z',
          author: {
            username: 'john',
            name: 'John Doe',
          },
        },
      ],
      pageNumber: 1,
      totalCount: 30,
      totalPages: 3,
    };

    moxios.stubRequest(new RegExp('/api/v1/posts/mock-post/comments.*'), {
      status: 200,
      response,
    });

    const { result, waitForNextUpdate } = renderHook(() => useCommentList('mock-post'), { wrapper: Providers });

    await waitForNextUpdate();

    expect(result.current).toMatchSnapshot();
  });
});
