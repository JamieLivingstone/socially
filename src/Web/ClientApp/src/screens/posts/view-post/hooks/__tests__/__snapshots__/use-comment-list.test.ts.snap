// Vitest Snapshot v1

exports[`useCommentList > returns paginated comment list 1`] = `
{
  "fetchNextPage": [Function],
  "hasNextPage": false,
  "isFetchingNextPage": false,
  "pages": [
    {
      "hasNextPage": false,
      "hasPreviousPage": false,
      "items": [
        {
          "author": {
            "name": "John Doe",
            "username": "john",
          },
          "createdAt": "2022-03-14T08:28:30.959Z",
          "id": 1,
          "message": "Comment one",
        },
      ],
      "pageNumber": 1,
      "totalCount": 30,
      "totalPages": 3,
    },
  ],
}
`;
